docs/history-service.md

```diff
@@ -0,0 +1,274 @@
+
+3. _Alternatively, if the request is a Worker reporting that the user workflow is blocked on a timer, then add a Timer Task in a History Service task queue; when that timer expires, create a Transfer Task representing the Workflow Task in Matching that is required to continue advancing the workflow._
+
+This is illustrated in the diagram below, which contains a typical sequence of events in the execution of the following workflow pseudocode (see [Workflow Lifecycle](./workflow-lifecycle.md) for more detailed scenarios):
+
+```javascript
+function myWorkflow() {
+  result = callActivity(myActivity);
+  workflowSleep(THIRTY_DAYS);
+  return result;
+}
+```
+
+```mermaid
+sequenceDiagram
+participant app as User Application
+participant worker as Worker
+participant handler as History Shard<br>RPC Handler
+participant qp as History Shard<br>Queue Processors
+participant matching as Matching Service
+app->>handler: StartWorkflowExecution
+note right of handler: State transition:<br>initialize Workflow Execution History<br>update Mutable State<br>create Transfer Task (Workflow Task)
+qp->>matching: enqueue Workflow Task
+worker->>handler: RespondWorkflowTaskCompleted(ScheduleActivity)
+note right of handler: State transition:<br>append to Workflow Execution History<br>update Mutable State<br>create Transfer Task (Activity Task)
+qp->>matching: enqueue Activity Task
+worker->>handler: RespondActivityTaskCompleted
+note right of handler: State transition:<br>append to Workflow Execution History<br>update Mutable State<br>create Transfer Task (Workflow Task)
+qp->>matching: enqueue Workflow Task
+worker->>handler: RespondWorkflowTaskCompleted(StartTimer)
+note right of handler: State transition:<br>append to Workflow Execution History<br>update Mutable State<br>create Timer Task
+qp->>qp: thirty days elapses<br><br>timer fired
+note right of handler: State transition:<br>append to Workflow Execution History<br>update Mutable State<br>create Transfer Task (Workflow Task)
+qp->>matching: enqueue Workflow Task
+```
+
+However, there are several requirements that make the above challenging to implement. In particular:
+
+- **Scaling**: the History service must be able to support arbitrarily many concurrent workflow executions. See [History shards](#history-shards).
+- **Consistency**: the Workflow and Activity tasks in Matching Service must be consistent with the workflow execution data in History Service. See [Consistency guarantees](#consistency-guarantees).
+- **Latency** of request handling and server-side workflow advance must be minimized.
+
+The rest of this document describes the implementation, which addresses these requirements.
+
+The History Service has many other responsibilities which are not covered here, for example: Shard Membership and Ownership, Replication and conflict resolution, Visibility, Archival, Workflow Versioning, Workflow Reset, Scheduled workflows.
```

Comment:
```suggestion
The History Service has many other responsibilities which are not covered here, for example: Shard Membership and Ownership, Replication and conflict resolution, Visibility, Archival, Workflow Reset.
```

history doesn't really do anything for versioning besides passing around some metadata. and it doesn't do anything special for scheduled workflows either (besides just that the scheduler workflow is a workflow, but it's not treated specially by history)